mlir/lib/Driver/Pipelines.cpp

```diff
@@ -67,43 +67,24 @@ void createQuantumCompilationPipeline(OpPassManager &pm)
-    gradient_buffer_options.opFilter.allowDialect<catalyst::gradient::GradientDialect>();
-    gradient_buffer_options.unknownTypeConverterFn =
-        [=](Value value, Attribute memorySpace,
-            const mlir::bufferization::BufferizationOptions &options) {
-            auto tensorType = cast<TensorType>(value.getType());
-            return bufferization::getMemRefTypeWithStaticIdentityLayout(tensorType, memorySpace);
-        };
-    pm.addPass(mlir::bufferization::createOneShotBufferizePass(gradient_buffer_options));
-    pm.addPass(mlir::createSCFBufferizePass());
     pm.addPass(mlir::createConvertTensorToLinalgPass());
     pm.addPass(mlir::createConvertElementwiseToLinalgPass());
-    pm.addPass(mlir::arith::createArithBufferizePass());
-    pm.addPass(mlir::bufferization::createEmptyTensorToAllocTensorPass());
-    pm.addNestedPass<mlir::func::FuncOp>(mlir::bufferization::createBufferizationBufferizePass());
-    pm.addNestedPass<mlir::func::FuncOp>(mlir::tensor::createTensorBufferizePass());
-    mlir::bufferization::OneShotBufferizationOptions catalyst_buffer_options;
-    catalyst_buffer_options.opFilter.allowDialect<catalyst::CatalystDialect>();
-    catalyst_buffer_options.unknownTypeConverterFn =
-        [=](Value value, Attribute memorySpace,
-            const mlir::bufferization::BufferizationOptions &options) {
-            auto tensorType = cast<TensorType>(value.getType());
-            return bufferization::getMemRefTypeWithStaticIdentityLayout(tensorType, memorySpace);
-        };
-    pm.addPass(mlir::bufferization::createOneShotBufferizePass(catalyst_buffer_options));
-    pm.addNestedPass<mlir::func::FuncOp>(mlir::createLinalgBufferizePass());
-    pm.addNestedPass<mlir::func::FuncOp>(mlir::tensor::createTensorBufferizePass());
-    mlir::bufferization::OneShotBufferizationOptions quantum_buffer_options;
-    quantum_buffer_options.opFilter.allowDialect<catalyst::quantum::QuantumDialect>();
-    pm.addPass(mlir::bufferization::createOneShotBufferizePass(quantum_buffer_options));
-    pm.addPass(mlir::func::createFuncBufferizePass());
-    pm.addNestedPass<mlir::func::FuncOp>(mlir::bufferization::createFinalizingBufferizePass());
+    pm.addPass(catalyst::createGradientPreprocessingPass());
+    pm.addPass(mlir::bufferization::createEmptyTensorEliminationPass());
+    ///////////
+    mlir::bufferization::OneShotBufferizationOptions options;
+    options.bufferizeFunctionBoundaries = true;
+    options.allowReturnAllocsFromLoops = true;
+    options.setFunctionBoundaryTypeConversion(
+        mlir::bufferization::LayoutMapOption::IdentityLayoutMap);
+    options.unknownTypeConverterFn = [=](Value value, Attribute memorySpace,
+                                         const mlir::bufferization::BufferizationOptions &options) {
+        auto tensorType = cast<TensorType>(value.getType());
+        return bufferization::getMemRefTypeWithStaticIdentityLayout(tensorType, memorySpace);
+    };
```

Comment:
Minor: In the textual pipeline we just use `unknown-type-conversion=identity-layout-map`, is there no equivalent C++ option like for the function boundary here?

Reply:
Alright no worries :)